pkg/registryclient/client.go

```diff
@@ -24,19 +32,28 @@ func Initialize(client kubernetes.Interface, namespace, serviceAccount string, i
 	kyvernoServiceAccount = serviceAccount
 	Secrets = imagePullSecrets
 
-	var kc authn.Keychain
-	kcOpts := &k8schain.Options{
+	kcOpts := &kauth.Options{
 		Namespace:          namespace,
 		ServiceAccountName: serviceAccount,
 		ImagePullSecrets:   imagePullSecrets,
 	}
 
-	kc, err := k8schain.New(context.Background(), client, *kcOpts)
+	k8s, err := kauth.New(context.Background(), client, *kcOpts)
 	if err != nil {
-		return errors.Wrap(err, "failed to initialize registry keychain")
+		return err
 	}
-
-	authn.DefaultKeychain = kc
+	// Note that we are using a multi keychain instead of k8schain.New since
+	// k8schain.New always includes the authn.DefaultKeychain in it's multikeychain
+	// see https://github.com/google/go-containerregistry/blob/50dfc2733d10b413e77747e0760d0b6136754a2f/pkg/authn/k8schain/k8schain.go#L49
+	// this means that if we don't manually initialize this keychain with our stored defaultKeyChain
+	// we would create an ever growing array of multikeychains each time we call update keychain
+	authn.DefaultKeychain = authn.NewMultiKeychain(
```

Comment:
Modifying the public var in another package is generally a pretty bad idea, and seems to be the root of the bug you're describing.

Can you instead pass around a keychain where it's needed, or else make your own public variable that you control yourself?

Reply:
The package could warn against this more forcefully, or even deprecate the var in favor of a method that returns an immutable value.

Reply:
@imjasonh I agree, I believe the original implementation was influenced by the advice at https://github.com/google/go-containerregistry/tree/main/pkg/authn/k8schain#using-the-keychain which was just set during initialization but was subsequently modified to handle secret updates https://github.com/kyverno/kyverno/pull/2798

I noticed this issue as I was using kyverno and proposed this fix. If the current fix isn't suitable we can look at the current calls to GGCR and pass in an explicit keychain option.

Reply:
Oh no. I'm going to fix that right now. 😄

Reply:
Is there a keychain in GGCR that resolves the passed pull secrets each time rather than fetching them during the creation of a new keychain?

Reply:
No, you'd need to create the keychain each time it's needed with the necessary secrets -- this is what Knative and Tekton do anyway. `Resolve` only takes the repo being resolved, not any other runtime args.

Knative:
- https://github.com/knative/serving/blob/492d71c96e7ada509f1cfeae5a8a04f1755fc9ef/pkg/reconciler/revision/resolve.go#L80

Tekton:
- https://github.com/tektoncd/pipeline/blob/6947f862d9f6aff3e9456562af6d4ff2ba3b98ea/pkg/pod/entrypoint_lookup_impl.go#L68

Reply:
Gotcha, I was hoping there might have been a keychain that stores the secret refs in the struct instead of the actual secret values and can fetch the appropriate references through the receiver for Resolve. If this is a common pattern or problem, might be worth providing a reference implementation in GGCR, otherwise we can construct it kyverno similar to knative/tekton.

Reply:
https://pkg.go.dev/github.com/sigstore/cosign/pkg/oci/remote#WithRemoteOptions will let you pass in your own keychain:

```go
kc, err := k8schain.New(...)
remoteOpts = append(remoteOpts, remote.WithAuthFromKeychain(kc))
```

which you can add here: https://github.com/kyverno/kyverno/blob/1af9e48b0dffe405c8a52938c78c710cf9ed6721/pkg/cosign/cosign.go#L50

Reply:
> Gotcha, I was hoping there might have been a keychain that stores the secret refs in the struct instead of the actual secret values and can fetch the appropriate references through the receiver for Resolve. If this is a common pattern or problem, might be worth providing a reference implementation in GGCR, otherwise we can construct it kyverno similar to knative/tekton.

That could be useful, but I'd be a bit worried about one struct holding potentially many clients' secrets in memory. It might be easy to confuse it into using the wrong ones. Seems better to just look them up each time they're needed. If you want to avoid API server traffic you could wrap `kubernetes.Client` interface with something that caches? But Tekton and Knative haven't hit an issue with it yet (as far as we know..... 😨 )

Reply:
@imjasonh updated to use our own keychain. Thanks for your help! :)